Permission guide (#394)
OntimizeWeb/docs: _docs/15c-permissions-example.md

---
title: Permissions example
permalink: /guide/permissions/example/
---

{% include base_path %} {% include toc %}

# Overview

In this section you gonna see the creation of an Ontimize application from cero using permissions on backend and frontend.

# Creation of the application

## First steps

To start the project we begin from a totally new Ontimize application. To see the complete example finished click the [following link](https://github.com/ontimize/ontimize-examples/tree/boot-web-permissions){:target="_blank"}.

## Configuring the backend

### Candidate entity creation

First we gonna do is to add an entity called Candidate and add it's service and rest controller as you can see [here](https://www.ontimize.com/xwiki/bin/view/Ontimize+Boot+Training/Creating+DAO%2C+Service%2C+Controller){:target="_blank"}. Then we gonna add records to our table candidate on the database.

> Once we got the previus steps we gonna create the user *candidate* on the database and  define his permissions and the admin permissions as you can see [here](https://ontimize.github.io/ontimize-boot/basics/security/){:target="_blank"}. In this example also you can see the tags that we use to secure the application and the methods. We use `@EnableAspectJAutoProxy(proxyTargetClass = false)` once on the `ServerAplication` class and `@Secured({ PermissionsProviderSecured.SECURED })` on all methods that you wanna protect.

Reviewer: ```suggestion
Once we got the previous steps we gonna create the user *candidate* on the database and  define his permissions and the admin permissions as you can see [here](https://ontimize.github.io/ontimize-boot/basics/security/){:target="_blank"}. In this example also you can see the tags that we use to secure the application and the methods. We use `@EnableAspectJAutoProxy(proxyTargetClass = false)` once on the `ServerAplication` class and `@Secured({ PermissionsProviderSecured.SECURED })` on all methods that you wanna protect.
```

---
title: Permissions example
permalink: /guide/permissions/example/
---

{% include base_path %} {% include toc %}

# Overview

In this section you are going to see the creation of an Ontimize application from cero using permissions on backend and frontend.

# Creation of the application

## First steps

> To start the project we begin from a totally new Ontimize application. To see the complete example finished click the [following link](https://github.com/ontimize/ontimize-examples/tree/boot-web-permissions){:target="_blank"}.

Reviewer: ```suggestion
This tutorial starts using an empty Ontimize application project and it is going to reach final configuration step by step. To see the complete example click the [following link](https://github.com/ontimize/ontimize-examples/tree/boot-web-permissions){:target="_blank"}.
```

## Configuring the backend

### Candidate entity creation

> First we are going to do is to add an entity called Candidate and add it's service and rest controller as you can see [here](https://www.ontimize.com/xwiki/bin/view/Ontimize+Boot+Training/Creating+DAO%2C+Service%2C+Controller){:target="_blank"}. Then we will add records to our table candidate on the database.

Reviewer: ```suggestion
First, we are going to add all stuff related with a Candidate: an entity, a service and the rest controller as you can see [here](https://www.ontimize.com/xwiki/bin/view/Ontimize+Boot+Training/Creating+DAO%2C+Service%2C+Controller){:target="_blank"}. Then, we will add records to our table candidate on the database.
```

> Once we got the previus steps we will create the user *candidate* on the database and  define his permissions and the admin permissions as you can see [here](https://ontimize.github.io/ontimize-boot/basics/security/){:target="_blank"}. In this example also you can see the tags that we use to secure the application and the methods. We use `@EnableAspectJAutoProxy(proxyTargetClass = false)` once on the `ServerAplication` class and `@Secured({ PermissionsProviderSecured.SECURED })` on all methods that you wanna protect.

Reviewer: ```suggestion
Next step is to create the user *candidate* on the database and  define his permissions and the admin permissions as you can see [here](https://ontimize.github.io/ontimize-boot/basics/security/){:target="_blank"}. In this example also you can see the tags that we use to secure the application and the methods. We use `@EnableAspectJAutoProxy(proxyTargetClass = false)` once on the `ServerAplication` class and `@Secured({ PermissionsProviderSecured.SECURED })` on all methods that you want to protect.
```

}
```

`com.ontimize.projectwiki.model.core.service.CandidateService`:
```java
@Override
@Secured({ PermissionsProviderSecured.SECURED })
public EntityResult candidateQuery(Map<String, Object> keyMap, List<String> attrList)
        throws OntimizeJEERuntimeException {
    return this.daoHelper.query(this.candidateDao, keyMap, attrList);
}
```

### Database records

Here are the rows that we added to our database on each table.

`TUSER table:`
![TUSER table]({{ base_path }}/images/permissions/database_tuser.png){: .align-center}

`TUSER_ROLE table`:
![TUSER table]({{ base_path }}/images/permissions/database_tuser_role.png){: .align-center}

`TROLE table:`
![TROLE table]({{ base_path }}/images/permissions/database_trole.png){: .align-center}

`TSERVER_PERMISSION table:`
![TSERVER_PERMISSION table]({{ base_path }}/images/permissions/database_tserver_permission.png){: .align-center}

`TROLE_SERVER_PERMISSION table:`
![TROLE_SERVER_PERMISSION table]({{ base_path }}/images/permissions/database_trole_server_permission.png){: .align-center}

As we can see on the last screenshot the only user that have access to the CRUD operations on the table is the `demo` user.

### Server permissions

> Now it's time to define the server permissions. In our application we got two users, Demo and Candidate. The only user that we want to have permission to access the candidates table it's the demo user so we will need to define it's permissions.

Reviewer: ```suggestion
Now it's time to define the server permissions. In our application we have two users, 'Demo' and 'Candidate'. The unique user that we want to have permission to access the candidates table is the 'Demo' user. So we will need to define it's permissions.
```

#### Creating the permission service

> Previous to define our permissions we need to create a rest controller with it's service to provide the permissions to the frontend.

Reviewer: ```suggestion
Previous to define our permissions, we need to create a rest controller with its service to provide the permissions to the frontend.
```

    static {
        try {
            CANDIDATE_PERMISSION = FileUtils.readFileToString(new File("projectwiki-model/src/main/resources/candidate_permissions.json"), StandardCharsets.UTF_8);
            DEMO_PERMISSION = FileUtils.readFileToString(new File("projectwiki-model/src/main/resources/demo_permissions.json"), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public EntityResult permissionQuery(Map<String, Object> keyMap, List<String> attrList)
            throws OntimizeJEERuntimeException {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        EntityResult e = new EntityResultMapImpl();
        Map<String, String> map = new HashMap<>();
        String role = authentication.getAuthorities().toArray()[0].toString();
        if (role.equals("candidate role")) {
            map.put("permission", PermissionService.CANDIDATE_PERMISSION);
        }
        else if (role.equals("admin")) {
            map.put("permission", PermissionService.DEMO_PERMISSION);
        }
        e.addRecord(map);
        return e;
    }
}
```

> #### Defining our server permissions

Reviewer: ```suggestion
#### Defining our client permissions
```

> Once we create the service we can add our JSON permission files. We can see more examples of this files [here](https://ontimizeweb.github.io/docs/v8/guide/permissions/#permissions-definition-example){:target="_blank"}.

Reviewer: ```suggestion
Client permissions are defined by a JSON file (you can find more detailed information  [here](https://ontimizeweb.github.io/docs/v8/guide/permissions/#permissions-definition-example){:target="_blank"}).
In this turorial, we are going to create a set of simple preconfigured JSON permission files. Into a real application, this permission object could be retrived from database or created dynamically by code depending on the user, the role, etc. We can see more examples of this files
```
